src/ui/mod.rs

```diff
@@ -23,7 +28,29 @@ impl<R: ResourcesProvider> Ui<R> {
             Command::Help => {
                 self.term.write_str(HELP_STR)?;
             }
+            Command::Login => {
+                self.login()?;
+            }
+            _ => {
+                self.term.write_str("Command not yet implemented\n")?;
+            }
         }
         Ok(())
     }
+
+    fn login(&mut self) -> Result<()> {
+        let login = Login {
+            username: self.prompt_username()?,
+            password: self.prompt_password()?,
+        };
+        service::login(&mut self.res, &login)
+    }
+    fn prompt_username(&mut self) -> Result<String> {
+        self.term.write_str(USERNAME_STR)?;
+        Ok(self.term.read_line()?)
+    }
+    fn prompt_password(&mut self) -> Result<String> {
+        self.term.write_str(PASSWORD_STR)?;
+        Ok(self.term.read_secure_line()?)
+    }
```

Comment:
Some functions can live here too, but maybe we should split things up into modules on the UI layer as well. The utility functions (`prompt_username`/`password`) only need `Term` so they at least can be moved.

Not definitive, but this is an option:

```rust
mod login; // private module, not visible outside ui
login::login(self)?;
// in login.rs:
use super::Ui;
use ...;
pub fn login(ui: &mut Ui<impl RP>) -> Result<()> {
    // private members of Ui are visible here, because this is a submodule
    // ...
}
fn prompt_username(&mut Term) -> Result<String> { ... }
fn prompt_password(&mut Term) -> Result<String> { ... }
```